src/storage/util.cpp

```diff
@@ -2,11 +2,64 @@
 // https://www.veriblock.org
 // Distributed under the MIT software license, see the accompanying
 // file LICENSE or http://www.opensource.org/licenses/mit-license.php.
+
+#include <algorithm>
 #include <veriblock/blockchain/alt_block_tree.hpp>
 #include <veriblock/storage/util.hpp>
 
 namespace altintegration {
 
+template <typename BlockTreeT>
+bool LoadTree(BlockTreeT& out,
+              BlockProvider<typename BlockTreeT::block_t>& provider,
+              ValidationState& state) {
+  using index_t = typename BlockTreeT::index_t;
+  using hash_t = typename index_t::hash_t;
+
+  auto it = provider.getBlockIterator();
+
+  if (!it->seek_start()) {
+    return state.Invalid("bad-iter", "cannot seek iterator to start");
+  }
+
+  std::vector<index_t> blocks;
+
+  for (; it->valid(); it->next()) {
+    blocks.push_back(it->value());
+  }
+
+  hash_t tip_hash;
+  if (!provider.getTipHash(tip_hash)) {
+    return state.Invalid("bad-provider", "cannot get tip");
+  }
+
+  if (!LoadTree(out, blocks, tip_hash, state)) {
+    return state.Invalid("bad-tree");
+  }
+
+  auto* t = out.getBestChain().tip();
+  VBK_ASSERT(t != nullptr);
+
+  return true;
+}
+
+bool LoadAllTrees(AltBlockTree& tree,
+                  BlockProvider<BtcBlock>& btc_provider,
+                  BlockProvider<VbkBlock>& vbk_provider,
+                  BlockProvider<AltBlock>& alt_provider,
```

Comment:
pass single provider here, then it will be just casted to specific types when passed by ref